feat: add support for application overloads (DHIS2-9092)

amcgee opened the pull request:
This PR adds a filter to the `dhis-web-portal` project which implements overloading of bundled core apps (`/dhis-web-xyz`) with an installed app of the same name (`/api/apps/xyz`)

To simplify this change and prevent issues with server discovery we are NOT changing app paths here.  The core apps are still served at the same paths (`/dhis-web-dashboard` etc.) but under the hood they will serve an installed `dashboard` app if it exists.  The path consolidation should be targeted for 2.36

TODO (in this PR):
- [x] If they exist, serve installed apps with names matching core apps from `/dhis-web-<name>`
- [x] 302 redirect `/api/apps/<name>` to `/dhis-web-<name>` for core apps
- [x] Confirm caching behavior
- [x] Hide override apps from modules list (apps menu)
- [x] Fix raw URL handling for overridden apps (i.e. `/dhis-web-scheduler` should redirect to `/dhis-web-scheduler/index.html`, does in bundled app but not when overridden) - also applies to any requested directory
- [x] Fix override name matching (uses label)
- [x] Expose `isBundledApp` property in `/api/apps` response
- [x] Expose correct `baseUrl` and `launchUrl` for bundled apps in API response
- [x] ~Expose list of bundled / overridable apps at `/api/apps/bundledAppNames`~ (removed)
- [x] Support filtering by `isBundledApp` in `/api/apps`

TODO (after this PR):
- Check performance - can/should we restrict the filter to a subset of routes in `web.xml` instead of `/*`?
- **security** Consider namespacing core apps or creating a new app type to prevent accidental overrides
- **security** Consider validating developer org of core apps for override?
- Deprecate `getModules.action` in favor of `/api/apps`
- Add support for "emergency" deletion of installed apps (maybe in the `/dhis-web-apps` project, or just through documentation) This is critical if AppManagement app is broken by a continuous release, which would be really nasty
- Look into excluding `manifest.webapp` from app cache.
- **In App Management App**
  - Support easy updating (maybe with an alert banner) of the App Management app itself, so we can defer these changes
  - Show bundled / core apps in their own section, warn of overrides
  - Fix App Icon in App Management app for overrides
  - Fix link to overridden apps in App Management

**NB** Includes some unrelated build automation additions, will remove those and put them in a separate PR

## Cache Behavior

`index.html` (including when served from `/`) appears to always avoid the cache, which was the existing behavior and works well here.  All other app files ARE cached, but since the vast majority include a hash in the filename (referenced from `index.html`) this should be fine.  The notable exception is `manifest.webapp` which could be stale - we should look into possibly excluding this from the cache as well**.  More testing is needed, but for the moment the app behavior after an app install / uninstall appears functional and consistent.

## Demo

A demo of this behavior (with an installed override of the Scheduler app) can be found at http://d2.winged.tech:8080 (admin:district)

## Incidental bugs fixed

Additional existing bugs in all Installed Apps fixed by this change:
- (in JCloudsAppStorageService.java) requests for non-existent files previously throw a 500 error instead of a 404
- (in JCloudsAppStorageService.java) requests pointing to directories (including the app route) would throw an error rather than serving `index.html` from that directory

stian-sandvold commented on dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/appmanager/DefaultAppManager.java:
```diff
@@ -308,7 +308,10 @@ public boolean isAccessible( App app )
     public boolean isAccessible( App app, User user )
     {
-        if ( user == null || user.getUserCredentials() == null || app == null || app.getName() == null )
+        if ( app == null || app.getName() == null || AppManager.BUNDLED_APPS.contains(app.getName()) ) {
```
```suggestion
        if ( app == null || app.getName() == null || AppManager.BUNDLED_APPS.contains( app.getName() ) ) {
```

stian-sandvold commented on dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/appmanager/JCloudsAppStorageService.java:
```diff
@@ -429,9 +430,21 @@ public Resource getAppResource( App app, String pageName )
+            
+            try {
+                res = locationManager.getFileForReading( filepath );
+            } catch (LocationManagerException e) {
```
```suggestion
            } catch ( LocationManagerException e ) {
```

stian-sandvold commented on dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/appmanager/JCloudsAppStorageService.java:
```diff
@@ -429,9 +430,21 @@ public Resource getAppResource( App app, String pageName )
-            if ( res.exists() )
+            if ( res.isDirectory() )
+            {
+                String indexPath = pageName.replaceAll("/+$", "") + "/index.html";
```
```suggestion
                String indexPath = pageName.replaceAll( "/+$", "" ) + "/index.html";
```

stian-sandvold commented on dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/appmanager/JCloudsAppStorageService.java:
```diff
@@ -429,9 +430,21 @@ public Resource getAppResource( App app, String pageName )
+            if ( res.isDirectory() )
+            {
+                String indexPath = pageName.replaceAll("/+$", "") + "/index.html";
+                log.info("Resource " + pageName + " (" + filepath + " is a directory, serving " + indexPath);
```
```suggestion
                log.info( "Resource " + pageName + " (" + filepath + " is a directory, serving " + indexPath );
```

stian-sandvold commented on dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/appmanager/JCloudsAppStorageService.java:
```diff
@@ -429,9 +430,21 @@ public Resource getAppResource( App app, String pageName )
+            {
+                String indexPath = pageName.replaceAll("/+$", "") + "/index.html";
+                log.info("Resource " + pageName + " (" + filepath + " is a directory, serving " + indexPath);
+                return getAppResource(app, indexPath);
```
```suggestion
                return getAppResource( app, indexPath );
```

stian-sandvold commented on dhis-2/dhis-web/dhis-web-commons/src/main/java/org/hisp/dhis/servlet/filter/AppOverrideFilter.java:
```diff
@@ -0,0 +1,177 @@
+    }
+
+    // From AppController.java (some duplication)
+    private void serveInstalledAppResource( App app, String resourcePath, HttpServletRequest request, HttpServletResponse response)
```
```suggestion
    private void serveInstalledAppResource( App app, String resourcePath, HttpServletRequest request, HttpServletResponse response )
```

stian-sandvold commented on dhis-2/dhis-web/dhis-web-commons/src/main/java/org/hisp/dhis/servlet/filter/AppOverrideFilter.java:
```diff
@@ -0,0 +1,177 @@
+        HttpServletResponse response = (HttpServletResponse) res;
+        String requestURI = request.getRequestURI();
+
+        Pattern p = Pattern.compile(APP_PATH_PATTERN);
```
```suggestion
        Pattern p = Pattern.compile( APP_PATH_PATTERN );
```

stian-sandvold commented on dhis-2/dhis-web/dhis-web-commons/src/main/java/org/hisp/dhis/servlet/filter/AppOverrideFilter.java:
```diff
@@ -0,0 +1,177 @@
+        String requestURI = request.getRequestURI();
+
+        Pattern p = Pattern.compile(APP_PATH_PATTERN);
+        Matcher m = p.matcher(requestURI);
```
```suggestion
        Matcher m = p.matcher( requestURI );
```

stian-sandvold commented on dhis-2/dhis-web/dhis-web-commons/src/main/java/org/hisp/dhis/servlet/filter/AppOverrideFilter.java:
```diff
@@ -0,0 +1,177 @@
+        Matcher m = p.matcher(requestURI);
+
+        if (m.find()) {
+            String namespace = m.group(0);
```
```suggestion
            String namespace = m.group( 0 );
```

stian-sandvold commented on dhis-2/dhis-web/dhis-web-commons/src/main/java/org/hisp/dhis/servlet/filter/AppOverrideFilter.java:
```diff
@@ -0,0 +1,177 @@
+
+        if (m.find()) {
+            String namespace = m.group(0);
+            String appName = m.group(1);
```
```suggestion
            String appName = m.group( 1 );
```

stian-sandvold commented on dhis-2/dhis-web/dhis-web-commons/src/main/java/org/hisp/dhis/servlet/filter/AppOverrideFilter.java:
```diff
@@ -0,0 +1,177 @@
+        if (m.find()) {
+            String namespace = m.group(0);
+            String appName = m.group(1);
+            String resourcePath = m.group(2);
```
```suggestion
            String resourcePath = m.group( 2 );
```

stian-sandvold commented on dhis-2/dhis-web/dhis-web-commons/src/main/java/org/hisp/dhis/servlet/filter/AppOverrideFilter.java:
```diff
@@ -0,0 +1,177 @@
+        Pattern p = Pattern.compile(APP_PATH_PATTERN);
+        Matcher m = p.matcher(requestURI);
+
+        if (m.find()) {
```
```suggestion
        if ( m.find() ) {
```

stian-sandvold commented on dhis-2/dhis-web/dhis-web-commons/src/main/java/org/hisp/dhis/servlet/filter/AppOverrideFilter.java:
```diff
@@ -0,0 +1,177 @@
+            String appName = m.group(1);
+            String resourcePath = m.group(2);
+
+            log.info("AppOverrideFilter :: Matched for URI " + requestURI);
```
```suggestion
            log.info( "AppOverrideFilter :: Matched for URI " + requestURI );
```

stian-sandvold commented on dhis-2/dhis-web/dhis-web-commons/src/main/java/org/hisp/dhis/servlet/filter/AppOverrideFilter.java:
```diff
@@ -0,0 +1,177 @@
+
+            log.info("AppOverrideFilter :: Matched for URI " + requestURI);
+
+            App app = appManager.getApp(appName);
```
```suggestion
            App app = appManager.getApp( appName );
```

amcgee commented on dhis-2/dhis-web/dhis-web-portal/src/main/webapp/WEB-INF/web.xml:
```diff
@@ -111,6 +111,12 @@
   </filter>
 
+  <filter>
+    <filter-name>org.hisp.dhis.servlet.filter.AppOverrideFilter</filter-name>
```
@larshelge should we mention the reason for this exception to the convention in a comment so future devs don't get confused and try to change it?  Can we add comments to `web.xml`?

larshelge replied:
That would be good :+1

amcgee replied:
I took a look at some other places using DelegatingFilterProxy and followed their example - can use `appOverrideFilter` here if using `@Component` decorator instead of adding the Bean to `beans.xml`, so made that change [here](https://github.com/dhis2/dhis2-core/pull/5977/commits/983a3884a71eff0482cf6b41d50ab21ecd2cf0d7).  No comment needed.

larshelge replied:
Sounds good.

stian-sandvold commented on dhis-2/dhis-web/dhis-web-commons/src/main/java/org/hisp/dhis/servlet/filter/AppOverrideFilter.java:
```diff
@@ -0,0 +1,177 @@
+
+            App app = appManager.getApp(appName);
+
+            if (app != null && app.getAppState() != AppStatus.DELETION_IN_PROGRESS) {
```
```suggestion
            if ( app != null && app.getAppState() != AppStatus.DELETION_IN_PROGRESS ) {
```

stian-sandvold commented on dhis-2/dhis-web/dhis-web-commons/src/main/java/org/hisp/dhis/servlet/filter/AppOverrideFilter.java:
```diff
@@ -0,0 +1,177 @@
+            App app = appManager.getApp(appName);
+
+            if (app != null && app.getAppState() != AppStatus.DELETION_IN_PROGRESS) {
+                log.info("AppOverrideFilter :: Overridden app " + appName + " found, serving override");
```
```suggestion
                log.info( "AppOverrideFilter :: Overridden app " + appName + " found, serving override" );
```

stian-sandvold commented on dhis-2/dhis-web/dhis-web-commons/src/main/java/org/hisp/dhis/servlet/filter/AppOverrideFilter.java:
```diff
@@ -0,0 +1,177 @@
+
+            if (app != null && app.getAppState() != AppStatus.DELETION_IN_PROGRESS) {
+                log.info("AppOverrideFilter :: Overridden app " + appName + " found, serving override");
+                serveInstalledAppResource(app, resourcePath, request, response);
```
```suggestion
                serveInstalledAppResource( app, resourcePath, request, response );
```

stian-sandvold commented on dhis-2/dhis-web/dhis-web-commons/src/main/java/org/hisp/dhis/servlet/filter/AppOverrideFilter.java:
```diff
@@ -0,0 +1,177 @@
+            if (app != null && app.getAppState() != AppStatus.DELETION_IN_PROGRESS) {
+                log.info("AppOverrideFilter :: Overridden app " + appName + " found, serving override");
+                serveInstalledAppResource(app, resourcePath, request, response);
+                // String newURI = "/api/apps/" + appName + requestURI.substring(namespace.length());
```
Comments with code

stian-sandvold commented on dhis-2/dhis-web/dhis-web-commons/src/main/java/org/hisp/dhis/servlet/filter/AppOverrideFilter.java:
```diff
@@ -0,0 +1,177 @@
+                // req.getRequestDispatcher(newURI).forward(req, res);
+                return;
+            } else {
+                log.info("AppOverrideFilter :: App " + appName + " not found, falling back to bundled app");
```
```suggestion
                log.info( "AppOverrideFilter :: App " + appName + " not found, falling back to bundled app" );
```

stian-sandvold commented on dhis-2/dhis-web/dhis-web-commons/src/main/java/org/hisp/dhis/servlet/filter/AppOverrideFilter.java:
```diff
@@ -0,0 +1,177 @@
+            }
+        }
+
+        chain.doFilter(req, res);
```
```suggestion
        chain.doFilter( req, res );
```

stian-sandvold reviewed (changes requested):
Some style issues, and some commented out code. The rest looks good to me :)

amcgee commented:
Thanks @stian-sandvold, fixed

amcgee commented on dhis-2/dhis-api/src/main/java/org/hisp/dhis/appmanager/AppManager.java:
```diff
@@ -36,13 +36,49 @@
 {
     String ID = AppManager.class.getName();
 
+    ImmutableSet<String> BUNDLED_APPS = ImmutableSet.of(
+        // Javascript apps
+        "app-management",
+        "cache-cleaner",
+        "capture",
+        "dashboard",
+        "data-administration",
+        "data-visualizer",
+        "data-quality",
+        "datastore",
+        "event-reports",
+        "event-visualizer",
+        "import-export",
+        "interpretation",
+        "maintenance",
+        "maps",
+        "menu-management",
+        "messaging",
+        "pivot",
+        "reports",
+        "scheduler",
+        "settings",
+        "tracker-capture",
+        "translations",
+        "usage-analytics",
+        "user",
+        "user-profile",
+        
+        // Struts apps
+        "approval",
+        "dataentry",
+        "maintenance-mobile"
+    );
```
Ideally this wouldn't duplicate the app cloning list in `dhis-web-apps`, but I don't think it would be trivial to refactor them at the moment.  Eventually this should go away when we move to the `/apps/xyz` app base

larshelge replied:
Agreed.

larshelge commented on dhis-2/dhis-web/dhis-web-commons/src/main/java/org/hisp/dhis/servlet/filter/AppOverrideFilter.java:
```diff
@@ -0,0 +1,175 @@
+        Matcher m = p.matcher( requestURI );
+
+        if ( m.find() ) {
+            String namespace = m.group( 0 );
```
Apply DHIS 2 code style.

amcgee replied:
Where is the code style documented?  Is there a lint tool?  @stian-sandvold [made the code-style recommendation](https://github.com/dhis2/dhis2-core/pull/5977#discussion_r475529098) (spaces around the 0) on this line

larshelge replied:
You can find info here: https://github.com/dhis2/wow-backend/blob/master/guides/code_formatting.md

larshelge replied:
Here specifically, brackets go on new lines.

amcgee replied:
Applied codeformatting

amcgee commented:
@stian-sandvold could you take another look at this?  @larshelge mentioned a code style issue, but I'm not sure what the proper style is.

larshelge commented on dhis-2/dhis-web/dhis-web-commons/src/main/java/org/hisp/dhis/servlet/filter/AppOverrideFilter.java:
```diff
@@ -0,0 +1,177 @@
+        if ( "manifest.webapp".equals( resourcePath ) )
+        {
+            // If request was for manifest.webapp, check for * and replace with host
+            if ( "*".equals( app.getActivities().getDhis().getHref() ) )
```
Perhaps some defensive null checks here to see if `activities` is actually defined to avoid NullPointerException, with some logging maybe.

amcgee replied:
Added

amcgee replied:
Also added to the existing code in `AppController.java` which was missing those checks (and copied here)

larshelge commented on dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/appmanager/JCloudsAppStorageService.java:
```diff
@@ -429,9 +430,21 @@ public Resource getAppResource( App app, String pageName )
-            File res = locationManager.getFileForReading( filepath );
+            File res;
+            
+            try {
```
Brackets on new lines.

amcgee replied:
Fied

amcgee commented:
This should be ready to go, however I'm unable to test the `@WebFilter` annotation change after the merge of the migration away from `web.xml` because `master` is still broken - will need to test it once `master` is fixed before merging this.

amcgee commented:
I've rebased this on the `2.35` branch, so this should be ready for review @larshelge @stian-sandvold @netroms 

I removed the `/api/apps/bundledAppNames` endpoint for now and made a few other minor improvements in the last commits

larshelge commented on dhis-2/dhis-web/dhis-web-commons/src/main/java/org/hisp/dhis/servlet/filter/AppOverrideFilter.java:
```diff
@@ -0,0 +1,178 @@
+    private ObjectMapper jsonMapper;
+
+    public static final String APP_PATH_PATTERN = "^/" + AppManager.BUNDLED_APP_PREFIX + "(" + String.join("|", AppManager.BUNDLED_APPS) + ")/(.*)";
+    // public static final String REDIRECT_APP_PATH_PATTERN = "^/" + AppController.RESOURCE_PATH + "-(" + String.join("|", AppManager.BUNDLED_APPS) + ")/(.*)"
```
Remove commented out code.
